cookbooks/cdo-varnish/libraries/helpers.rb

```diff
@@ -0,0 +1,135 @@
+# Various Ruby helper methods to generate Varnish VCL from the base configuration.
+
+# Basic regex matcher for the optional query part of a URL.
+QUERY_REGEX = "(\\?.*)?$"
+
+# Takes an array of path-patterns as input, validating and normalizing
+# them for use within a Varnish regular expression.
+# Returns an array of extension patterns and an array of path-prefixed patterns.
+def normalize_paths(paths)
+  paths.map(&:dup).partition do |path|
+    # Strip leading slash
+    path.gsub!(/^\//,'')
+    is_extension = path[0] == '*'
+    if !valid_path?(path) ||
+      (is_extension && path.match('/'))
+      raise ArgumentError.new("Invalid path: #{path}")
+    end
+    # Strip leading/trailing wildcard
+    path.gsub!(/(^\*|\*$)/,'')
+    # Escape some special characters
+    path.gsub!(/[.+$"]/){|s| '\\' + s}
+    is_extension
+  end
+end
+
+# The maximum length of a path pattern is 255 characters.
+# The value can contain any of the following characters:
+# A-Z, a-z (case sensitive, so the path pattern /*.jpg doesn't apply to the file /LOGO.JPG.)
+# 0-9
+# _ - . $ / ~ " ' @ : +
+#
+# The following characters are allowed in CloudFront path patterns, but
+# are not allowed in our configuration format to reduce complexity:
+# * (exactly one wildcard required, either at the start or end of the path)
+# ? (No 1-character wildcards allowed)
+# &, passed and returned as &amp;
+def valid_path?(path)
+  # Maximum length
+  return false if path.length > 255
+  # Valid characters allowed
+  char = /[A-Za-z0-9_\-.$\/~"'@:+]/
```

Comment:
This isn't the full set of valid characters allowed in an URL.  If you wanted you could tighten the documentation to be more restrictive if we don't care. 

http://stackoverflow.com/questions/1547899/which-characters-make-a-url-invalid
"n general URIs as defined by RFC 3986 (see Section 2: Characters) may contain any of the following characters: ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~:/?#[]@!$&'()*+,;="